Migrate skill creation service to angular

srijanreddy98 opened the pull request:
## Overview
<!--
READ ME FIRST: Please answer *both* questions below and check off every point from the essential checklist!
-->
This pr aims to migrate $rootscope.loadingMessage by providing a service for it. This pr also migrates skill creation service to angular 8.
1. This PR fixes or fixes part of #[fill_in_number_here].
2. This PR does the following: [Explain here what your PR does and why]
This pr creates a file called CommonEventsService which helps to migrate $rootScope. In the next comment, I have mentioned a possible plan of action.
## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise, this will lead to review delays
- Never force push. If you do, your PR will be closed.

srijanreddy98 committed:
Migrate skill creation service to angular

srijanreddy98 commented:
# Migrating RootScope

When migrating angularjs services to Angular you would come across $rootScope. There is no equivalent of $rootScope in Angular. We have to migrate by creating an Anguar Service called CommonEventsService.

In the services present in code base there are 77 instances of $rootScope spread across 33 services. They are:

 1. \$rootScope.$broadcast (57 instances)
 2. \$rootScope.loadingMessage (13 instances)
 3. \$rootScope.$watch (3 instances)
 4. \$rootScope.$apply (3 instances)
 5. \$rootScope.$on (1 instance)

## \$broadcast
This is used to broadcast that a certain event has happened. You will find it in the code this way:
`$rootScope.$broadcast(some-event)`
There will be someother place in the code listening to this event. So we have to notify it. To do this in Angular we will use RxJS subjects. An RxJS Subject is a special type of Observable that allows values to be multicasted to many Observers. When you see a $broadcast(event), come over to commonEvents service and check if there is a function in the service which has a camelCase name of the event. If you find that function then use it otherwise follow these steps:

 - Create a new subject
	```
	private _eventA =  new  Subject<any>();
	```
 - Create a getter for this event so that it can used for \$rootScope.$on instances
	```
	onEventA():Subject {
		return this._event;
	}
	```
	
 - Create a broadcast function
	```
	 broadcastEventA():void {
		 this._eventA.next();
	 }
	```
	There is however another way to the same, things will be very generic and a lot of less code will be used. We can just create one function which handles all events with losing type safety being one of the caveats.
	```
	emitEvent(eventName,  ...args):  void  {
		this[eventName].emit(...args);
	}
	```
## LoadingMessage
In order to migrate \$rootScope.loadingMessage, we can create an RxJS subject called _loadingMessage in the Angular CommonEventsService. 
 - Create a new loadingMessage subject
	```
	private _loadingMessage: Subject = new Subject<string>();
	```
- A getter function for this
	```
	getLoadingMessage(): Subject {
		return this._loadingMessage;
	}
	```
- A an emitter function for this
	```	
	setLoadingMessage(message): void {
		this._loadingMessage.next(message);
	```
- How to use these subjects? Please look at the end of the docs.

## $Watch
We have three instances of \$rootScope.$watch. Each of these instances are used for watching the urls. This can now be done in Angular using the Angular router. Here is an example:
```
import  {  Router, NavigationEnd  }  from  '@angular/router';
...
export class SomeService {
	...
	constructor(
		...
		private  router:  Router,
		...
	){}
	...
	this.router.events.subscribe(event  =>  {
		if (event  instanceof  NavigationEnd) {
			// Here event.url is the final url
		}
	});
	}
	...
}
```
## \$apply
\$rootScope.$apply() is no longer required in Angular. It can be removed from the code.

## \$on

\$rootScope.$on can be mingrated in the following way. 
In the code \$rootScope.$on will be pretty much used this way:
```
$rootScope.$on(someEvent, function() {
	...
});
```
This can be changed to 
```
this.commonEventsService.onSomeEvent().subscribe(
	() => { // This is the same function as above, just using fat arrow syntax
		...
	}
);
```

**This can be used when we are migrating all the services at once which is not the case for us. We are doing it incrementally. We need to make sure the old way works too. A small modification can help us achieve that.**

To the CommonEvents service constructor, we add the following:
```
private  _rootScope;
constructor()  {
	const  injector = angular.element(document.getElementsByTagName('html')[0]).injector();
	this._rootScope  =  injector.get('$rootScope');
}
```
There is another way for getting the $injector that didn't work. Just mentioning here for the sake of documenting differnet methods tried.
```
const  injector  =  angular.injector(['ng',  'oppia']);
```

To each of the broadcast function we will add `this._rootScope.$broadcast(eventA)`,  so it will look like:
```
broadcastEventA():void {
	this._eventA.next();
	this._rootScope.$broadcast(eventA);
}
```
and to each of onEvent function we will add `this._rootScope.$on(eventA,  listener)`
```
onEventA(listener):Subject {
	this._rootScope.$on(eventA,  listener)
	return this._event;
}
```
You have probably noticed that initially the function didn't have a listner argument but now it has. So once we have migrated all we will remove this so, whereever you have used this add a comment.  `//TODO(issueNo): Remove listner argument when migrated to Angular`
Also note that using normal function() notation will remove the context of this, it is better to pass as an arrow function, that way, this will still refer to the place where you sent the function from otherwise you will have to use .bind(this).

## RxJS subjects
RxJS subjects are what we are using to emulate $rootScope. So it would be helpful if you gained some insight into this.
```
import  {  Subject  } from 'rxjs';

const subject =  new  Subject<string>();

subject.subscribe((v)  => console.log(`observerA: ${v}`));
subject.subscribe((v)  => console.log(`observerB: ${v}`));

subject.next('Hello');
subject.next('World');

// Output will be
// observerA: Hello
// observerB: Hello
// observerA: World
// observerB: World
```
If you want to read more about this you can have a look at subjects [here](https://rxjs-dev.firebaseapp.com/guide/subject)

srijanreddy98 committed:
Changes to warning-loader.directive.ts

srijanreddy98 committed:
Fix failing travis test

srijanreddy98 committed:
Fix failing travis test

bansalnitish commented on core/templates/services/common-events.service.ts:
```diff
@@ -0,0 +1,59 @@
+export class CommonEventsService {
+  // TODO(srijanreddy98): Remove the _rootScope once the project is migrated.
+  private _rootScope;
+  private _loadingMessage = new Subject<string>();
```
One concern here is why we are using `Subject` and not `Observable`?

srijanreddy98 replied:
I may be wrong here but observables by default are cold. Instead of making them hot (so that they can be broadcast to multiple subscribers), I have just used subjects which means a lot less code. As far as I know, one can't push new values to an observable (outside of where you create it). Basically a subject is a hot observable with the ability to pump in new values.

My views are based off this medium article (https://medium.com/@benlesh/on-the-subject-of-subjects-in-rxjs-2b08b7198b93). It is written by Ben Lesh, who is an RxJS lead.
and also a youtube video(https://www.youtube.com/watch?v=ewcoEYS85Co)

bansalnitish replied:
I didn't get the word "hot" and "cold". 

> As far as I know, one can't push new values to an observable (outside of where you create it). Basically a subject is a hot observable with the ability to pump in new values.

Also, could you explain this may be with the help of an example?

srijanreddy98 replied:
Writing about hot and cold observables could be a bit long, I would refer this query to an excellent read from the makers of RxJS
https://medium.com/@benlesh/hot-vs-cold-observables-f8094ed53339

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks! Just a general note: Instead of mimicking the behavior of the `$rootScope` shouldn!t we try to get rid of the usages since it is probably not a smart way of doing these things?

kevinlee12 reviewed (approved):
lgtm for code owner files.

Showtim3 commented:
@srijanreddy98 @bansalnitish Hey guys, there are $rootscope in a lot of directives as well, will they work if we migrate the $rootscope in services only?

srijanreddy98 commented:
> LGTM from code owner's perspective. Thanks! Just a general note: Instead of mimicking the behavior of the `$rootScope` shouldn!t we try to get rid of the usages since it is probably not a smart way of doing these things?

@vojtechjelinek True, it is not the smart way to do it. But there are certain use cases that require a "global" service. For example, If a component wants to show the loading screen then it would require a service to notify the loader to show the loading screen. If the entire project were in Angular, we could make a shared components module and import the loader component from there and use it. But this can only be done if our entire codebase is in Angular. 

So basically any idea one can think of to mostly cannot be implemented at the moment because the whole project hasn't been migrated yet. Once the project is fully migrated, we can try to limit certain "events" to stay within their respective modules.

srijanreddy98 commented:
> @srijanreddy98 @bansalnitish Hey guys, there are $rootscope in a lot of directives as well, will they work if we migrate the $rootscope in services only?

@Showtim3 Yes, they will. This service is created in such a way that it will work with both the old code and the new code.

jameesjohn reviewed (approved):
LGTM! for code owner files

bansalnitish commented:
> Instead of mimicking the behavior of the $rootScope shouldn!t we try to get rid of the usages since it is probably not a smart way of doing these things?

Me and @srijanreddy98 will discuss this point offline once and update the thread.

vojtechjelinek commented:
@bansalnitish @srijanreddy98 Thanks, this makes sense, just wanted to make sure that we are following good practices.

nishantwrp commented on core/templates/components/entity-creation-services/skill-creation.service.ts:
```diff
@@ -16,45 +16,52 @@
+    this.commonEventsService.setLoadingMessage('Creating skill');
+    this.skillCreationBackendApiService.createSkill(
+      description, rubrics, explanation, linkedTopicIds
+    ).then((response: any) => {
```
Hi @srijanreddy98,
Can't the exact type for `response` be declared?

srijanreddy98 replied:
Done!

srijanreddy98 committed:
Small Changes

srijanreddy98 committed:
Small Changes

srijanreddy98 commented:
# Common Events Service
Common events service is a service that we plan to use in the codebase to notify of certain events happening. 

## Is the rootScope equivalent in Angular2+?
On a quick look, it can seem that this provides the $broadcast and $on features that the $rootScope used to provide. However, this is very different from the $rootScope. When using $broadcast the event is broadcasted across the code base, everywhere. Only the $on which matches reacts to it, on the other hand in the common events service, we use RxJS subjects. Subjects keep an internal list of observers and only send the data to them. So this is not a global broadcast of message, it is more like one to one communication. However, there are certain events that can be handled within a module. Like events pertaining to a certain module. We can have a module-specific service to handle such events and one common service to handle events across the module.

Please note that common events service is not a replacement for $rootScope or the $rootScope of Angular2+, it just a service to help implement the existing in code in a new way. There are also instances of \$rootScope.$watch which are looking at the URL, we will be using the Angular router to handle that.

## What else will it be used for?

The common events service, in theory, can also hold some state data of the current application, like if it is running in dev_mode or not. But ideally, a .env file used for such purposes.

## Why is there a rootScope in the service?

In the pr [#9048](https://github.com/oppia/oppia/pull/9048/), the service is acessing the $rootScope. This is just a temporary measure to handle this migration incrementally. This wouldn't be required if all of the files were migrated at once. But it is not possible to do so in our current code base which has around 70 services left and over 300 directives to be migrated. So we have to do it incrementally without breaking the exisiting code, and hence we have to use $rootScope initially. But once the migration is over it can be safely removed and we will have no more instances of the $rootScope.

bansalnitish commented:
@vojtechjelinek @seanlip could you please check the latest comment by @srijanreddy98 and let us know your concerns?

seanlip commented:
Thanks @srijanreddy98 and @bansalnitish. I am a bit confused about the comment -- is it meant to be docs for users or is it meant to be just a comment for this PR? Assuming the former, I took an edit pass and left some concerns inline -- PTAL.

I'm also not really sure about the approach taken for the last question. Why don't you implement common-events.service.ts first and then transition other services to use it directly? You can do that message by message, so I don't see why it needs to block on the full migration.

-------------------------------- 

# Common Events Service
The CommonEventsService is a service that we plan to use in the codebase to provide notifications when certain events happen.

## Is this the equivalent of rootScope in Angular 2+?
No. In AngularJS, when using $broadcast, the event is broadcasted across the code base, everywhere, and only the $on which matches reacts to it. However, in CommonEventsService, we use RxJS subjects. Each subject keeps an internal list of observers, and only sends the data to them. So, this is not a global broadcast of the message, it is more like one-to-one communication. Please note that we are trying to move all message-passing within the codebase to use the CommonEventsService instead of $rootScope. 

Note that there are certain events that can be handled within a module (e.g. module-specific events). We can have a module-specific service to handle such events, and use a common service to handle events across the module. **[I think you need more detail here -- I tidied up this paragraph but I don't understand it. Maybe provide some specific examples?]**

There are also instances of \$rootScope.$watch which are looking at the URL, we will be using the Angular router to handle that. **[Not sure what this sentence has to do with this question.]**

## What else will this service be used for?

Nothing. In theory, the CommonEventsService can also hold some state data of the current application, like whether or not it is running in dev_mode. But, ideally, we would use an .env file for such purposes. **[This seems off to me. I would have expected something like constants.ts instead?]**

## Why is there a rootScope in the service?

In the pr [#9048](https://github.com/oppia/oppia/pull/9048/), the service is accessing the $rootScope. This is just a temporary measure to handle this migration incrementally. This wouldn't be required if all of the files were migrated at once. But it is not possible to do so in our current code base which has around 70 services left and over 300 directives to be migrated. So we have to do it incrementally without breaking the existing code, and hence we have to use $rootScope initially. But once the migration is over it can be safely removed and we will have no more instances of the $rootScope.

srijanreddy98 commented:
Thanks @seanlip for the review. This was meant to both as a comment and also (if this pr goes through) for the docs. And also I wanted to document the reasons behind certain choices being made.
> I'm also not really sure about the approach taken for the last question. Why don't you implement common-events.service.ts first and then transition other services to use it directly? You can do that message by message, so I don't see why it needs to block on the full migration.

I am not really sure if I get your question right, please correct me if this doesn't answer your question:
Currently in angularjs, `$rootScope.$broadcast('event')` is only caught by `$rootScope.$on('event', ...)`. If we migrate a file which contains $rootScope.$broadcast, it will no longer be received by the $on('event') (in angularjs). This will break the existing code. One way to do it will be to migrate these pairs together, but since we are migrating services first, and mostly $on instances are in directives, it won't be feasible. So until the migration is complete, I am notifying both, angularjs and angular parts of the code of the events happening.

> [I think you need more detail here -- I tidied up this paragraph but I don't understand it. Maybe provide some specific examples?]
That's a wording problem on my side. I meant to write: and use a common service to handle events across different modules.

Taking the example of EVENT_AUTOPLAY_AUDIO, this occurs in five places across three files.

File | Rootscope. | Event Name
-- | -- | --
./oppia/core/templates/pages/exploration-player-page/learner-experience/conversation-skin.directive.ts | broadcast | EVENT_AUTOPLAY_AUDIO
./oppia/core/templates/pages/exploration-player-page/learner-experience/conversation-skin.directive.ts | broadcast | EVENT_AUTOPLAY_AUDIO
./oppia/core/templates/pages/exploration-player-page/layout-directives/audio-bar.directive.ts | on | EVENT_AUTOPLAY_AUDIO
./oppia/core/templates/pages/exploration-player-page/services/hint-and-solution-modal.service.ts | broadcast | EVENT_AUTOPLAY_AUDIO
./oppia/core/templates/pages/exploration-player-page/services/hint-and-solution-modal.service.ts | broadcast | EVENT_AUTOPLAY_AUDIO

As all of this is within the `exploration-player-page.module` we can have something like `exploration-player-page-events.service.ts`, where we will add the code for broadcast and on. This way the code will be more organized.

> There are also instances of $rootScope.$watch which are looking at the URL, we will be using the Angular router to handle that. [Not sure what this sentence has to do with this question.]

Yeah, I will remove it, it was just to show that if common events service were rootScope, we wouldn't be using the angular router to migrate it.

> [This seems off to me. I would have expected something like constants.ts instead?]

Sorry, my bad. constants.ts is more appropriate.

srijanreddy98 commented:
@nishantwrp I have made the changes you requested, could you ptal?

seanlip commented:
Thank you for the clear explanation, @srijanreddy98! I really appreciate it.

I was indeed thinking of migrating the stuff in parallel for each event name (e.g. everything related to EVENT_AUTOPLAY_AUDIO, or 'event', or whatever the string is). This would include both $broadcast and $on for each event.

I would suggest, instead of your proposal to have a temporary $rootScope, doing the following instead:
- Move $on to some service instead so it's not in a directive
- Migrate event-by-event as mentioned above

Would that work? That way, you avoid introducing new bad patterns in the codebase.

vojtechjelinek commented:
Hi, thanks for the explanation, I agree with Sean that unifying the events into the services makes sense and will also make more clear from which places we call which events.

bansalnitish commented:
One question for you @seanlip: Migrating events by events would involving migrating those 3 directives and 2 services in one shot (for the above example, there could be more work involved). Should we go ahead with this?

seanlip commented:
Yes. I think that's fine as long as there is a clearly-defined and
documented methodology that is easy for contributors to follow.

>

bansalnitish commented:
We cannot migrate any directive. Directives too have to migrated in the dependency order so I think this is something not possible right now.

srijanreddy98 commented:
> We cannot migrate any directive. Directives too have to migrated in the dependency order so I think this is something not possible right now.

@seanlip the quoted comment is true. However, we can try your suggestion in a differnt way. I am trying to access the common events service in angularjs code and subscribe to the event subject, which would obviate the need of migrating the directives. But at the moment, it just refuses to subscribe to the subject in the angularjs code. I will try to make this work though.
## What I am trying to do:
### For $broadcast of eventA
 - Create a new subject
	```
	private _eventA =  new  Subject<any>();
	```
 - Create a getter for this event so that it can used for \$rootScope.$on instances
	```
	onEventA():Subject {
		return this._event;
	}
	```
	
 - Create a broadcast function
	```
	 broadcastEventA():void {
		 this._eventA.next();
	 }
	```
### $on of events (trying this in angularjs)
```
CommonEventsService.onEventA().subscribe(
	() => { 
		...
	}
);
```

Explanation: 
The onEventA() returns a subject, which is subscribed here. (the problem being it is not subscribing to the subject in angularjs code)
As for the broadcast, it will work perfectly fine in any code (i.e., angular or angularjs)
So whenever the function `broadcastEventA()` is called it will trigger the subject which will inturn notify all receivers, i.e. the places where it was subscribed.

seanlip commented:
Actually, wait. I might have misunderstood your comment. I was talking
about moving the broadcast/on/emit to services, not migrating directives.

>
>

srijanreddy98 commented:
> Actually, wait. I might have misunderstood your comment. I was talking about moving the broadcast/on/emit to services, not migrating directives.

@seanlip I am not really sure if moving the subscriptions($on part) to services would be a great idea. In Angular (not really sure about angularjs) the lifecycle of service is, it is constructed when the first component/directive/service/Pipe who needs is constructed. And it is destroyed only when the module they are declared in is destroyed(most of the time when the page is unloaded). So these subscriptions would survive even if the component using that subscription is destroyed, which is not a good thing to have.

bansalnitish commented:
I agree with what Sean's said -- that is something we can do. Remove all `$rootScope` things from directive and keep them in a common service(per module) and then use that service in the directive. 

@srijanreddy98 I didn't get exactly what would happen using services in place of directives to call the $rootScope things.

srijanreddy98 commented:
> @srijanreddy98 I didn't get exactly what would happen using services in place of directives to call the $rootScope things.

@bansalnitish 
So basically everything in Angular has a life cycle. When something is created, when something is destroyed, etc.
I will take a simple example here to make my case:
If there is an element with a ngIf directive, when the condition is evaluated to be false, angular destroys the component (one can check using ngOnDestroy). When this happens everything of the component is destroyed (so basically any subscriptions will be unsubscribed).
Service has a different lifecycle (mentioned in my previous comment) as they don't appear in the DOM.

In a case where the subscription of an event (for mimicking $on) was in a directive, it would be unsubscribed (i.e. won't listen to the event broadcast) when *ngIf is false.
However, if the same subscription was in a service, it would not be destroyed when ngIf was evaluated to be false (which is not the behavior we want). It would still be subscribed, and a copy of the component will be still in memory (bad memory management). When the ngIf evaluates to be true, a new instance of the component will be created and it will have a new subscription (so there will be two subscriptions active). Hence moving this to a service is not a good idea in my opinion.

seanlip commented:
Hi @srijanreddy98, I'm really glad you are considering the lifecycle of objects. That is really good, and thanks for doing this!

Here is a possible way to address this issue. In general, complex logic should not be in directives anyway. So a lot of the stuff is currently misplaced and should really be in a service. However, this needs to be done case-by-case in order to do it safely -- you'll need to think about what's actually going on with the event and what needs to trigger based on the message. Directives should generally just be for UI elements and not logic, but usually messages trigger some sort of logic.

So, you'll need to look into what the actual impact of receiving the event is, on a case-by-case basis. What is triggered by the $on callback? Usually it's some kind of logic, in which case it makes sense for all of this code to be in a service anyway. If it's some UI element then what you can do is to set a watch in the directive that listens to a variable in the service saying whether there's a pending request that still needs to be handled. But I think this latter case is likely to be rarer.

Does that help? The point I'm trying to make is that most of this code was probably not supposed to be in directives in the first place.

srijanreddy98 commented:
Hi @seanlip, 
+cc @bansalnitish, @vojtechjelinek 
I tried to refactor **EVENT_SKILL_INITIALIZED**. It has three occurrences. One broadcast and two on.
File | Rootscope. | Event Name
-- | -- | --
skill-editor-navbar.directive.ts | on | EVENT_SKILL_INITIALIZED
skill-questions-tab.directive.ts | on | EVENT_SKILL_INITIALIZED
skill-editor-state.service.ts | broadcast | EVENT_SKILL_INITIALIZED

This broadcast is triggered in the function `_setSkill` which is only called from the function `_updateSkill`. `_updateSkill` is called from two functions: `loadSkill` and `saveSkill`.

### loadSkill:

loadSkill function is called in two places:

1. When the skill-editor-page.controller.ts is initialized.
2. When changes are discarded (in skill-editor-navbar.directive).

### saveSkill:
1. saveSkill is only called in skill-editor-navbar.directive when the skill is saved.

Due to my poor angularjs skills (never done angularjs, only Angular),  I can't think of a way to move this into services without rootScope in use. What is confusing me, even more, is in skill-editor-navbar.directive there is a function which is linked to three events.
```
$scope.$on(EVENT_SKILL_INITIALIZED, _validateSkill);
$scope.$on(EVENT_SKILL_REINITIALIZED, _validateSkill);
$scope.$on(
    EVENT_UNDO_REDO_SERVICE_CHANGE_APPLIED, _validateSkill);
```
I am getting really confused about how to move them out from directives. As you mentioned in the previous comment that directives shouldn't have much app logic which is true for angularjs. But in these instances, it looks like the logic does belong in these files (I could be wrong here).

So basically I can think of a lot of ways but only in Angular. I don't think it is possible to do it or I am not able to see a clear way of refactoring in angularjs due to my lack of knowledge in angularjs, so, if it is possible that someone could give me more hints or a small example, that would really help me a lot.

seanlip commented:
In the given page, is there only one skill object that exists at a time? If so, you could have a service that stores that skill object and exposes an API that the directive can use to retrieve it (or retrieve its validation errors, etc.). Then _validateSkill would be in that service and that service can receive broadcast events. Would that work?

(Note that this works only if there is only one skill object in this context. It would be trickier if there are multiple skill objects that the page relates to.)

srijanreddy98 commented:
@seanlip, but wouldn't that be a problem, if skill-editor-navbar.directive is not loaded, the $on in the service would still receive the broadcast and it would still be validating the skill even if it is not required.

Also would like to point out that using $watch will increase the complexity of migration.

srijanreddy98 commented:
secondly, storing app-state data in services could run into a tricky problem. Currently, we are migrating angular services and adding some of them to Upgraded Services. 
Let us say we have two or more directives accessing the service (directive_A and directive_B).
Directive A has been migrated to angular 8 while directive_B is still angular js.
Directive B uses the service instance created in UpgradedService.ts while directive_A is provided with the instance Angular makes, so they will be accessing different instances of the service hence will have different data.
So, if directive A makes changes to skill data, it will not be reflected/ accessible to directive B.

We haven't faced this problem so far because we are using mostly HTTP calls, which is independent. (i.e. doesn't matter which instance is calling it, will return what is expected from the server)

seanlip commented:
> @seanlip, but wouldn't that be a problem, if skill-editor-navbar.directive is not loaded, the $on in the service would still receive the broadcast and it would still be validating the skill even if it is not required.

No, the whole point of this is for all the broadcast/on logic to stay in services. So whether the directive has loaded or not becomes irrelevant in this scenario . The trigger for validation is simply that the event gets fired.

> Also would like to point out that using $watch will increase the complexity of migration.

I don't understand where you need to use $watch? You can just return a reference to the skill stored in the service.

> secondly, storing app-state data in services could run into a tricky problem. Currently, we are migrating angular services and adding some of them to Upgraded Services.

We already store app-state data in services, so I think this is an issue that may need to be solved in general anyway.

srijanreddy98 commented:
> No, the whole point of this is for all the broadcast/on logic to stay in services. So whether the directive has loaded or not becomes irrelevant in this scenario. The trigger for validation is simply that the event gets fired.

The EVENT_SKILL_INITIALIZED triggers two functions, one in skill-questions-tab.directive.ts, the other in skill-editor-navbar.directive.ts. Only one them exists at a time. Keeping them in services will trigger both the functions.

Wouldn't it be redundant to trigger both these functions when it is quite evident that only one needs to be executed at a time?

The function triggered in skill-questions-tab.directive
```
var _init = function() {
            $scope.skill = SkillEditorStateService.getSkill();
            $scope.getQuestionSummariesAsync =
              QuestionsListService.getQuestionSummariesAsync;
            $scope.getGroupedSkillSummaries =
              SkillEditorStateService.getGroupedSkillSummaries;
            $scope.isLastQuestionBatch =
             QuestionsListService.isLastQuestionBatch;
            $scope.skillIdToRubricsObject = {};
            $scope.skillIdToRubricsObject[$scope.skill.getId()] =
              $scope.skill.getRubrics();
          };
```
The function is skill-editor-navbar.directive.ts
```
var _validateSkill = function() {
            $scope.validationIssues = $scope.skill.getValidationIssues();
          };
```

seanlip commented:
This is not a cookie-cutter refactor. You really need to analyze the code and trace a bit to understand why it is that way, then you can find a better way to rewrite it.

Firstly, I don't understand what you mean by "Only one of them exists at a time.". In skill-editor-page.controller.ts, both these files are imported, which means they exist simultaneously on the same page.

Secondly, let's look at the navbar. It is listening to changes in the skill ID. When that happens, it wants to re-run the validation so that it can show the correct validation errors. 

Thirdly, let's look at the questions tab. The 2nd-4th statements in _init() can be left as-is. Everything else is all about just getting the updated skill correctly when the "initialize/reinitialize" event is fired, so that the skill and the rubrics can be derived from that.

So, one way to proceed is as follows. In the questions tab, bind $scope.skill directly to the current skill in SkillEditorStateService, in such a way that when the value in the latter changes, $scope.skill changes. Then, in the questions tab, add a watch to $scope.skill.id and run a bunch of things when that ID changes. Similarly for the navbar. Then you won't even need a broadcast/on at all, which is even better.

(The reason you end up not needing a broadcast/on at all here is that the natural "service" to transfer the logic to for both of these directives is SkillEditorStateService, and a service need not broadcast to itself.)

srijanreddy98 commented:
> Firstly, I don't understand what you mean by "Only one of them exists at a time.". In skill-editor-page.controller.ts, both these files are imported, which means they exist simultaneously on the same page.

Due to an ng-if directive present, the skill questions tab is not in dom/ loaded until it is clicked on the navbar.
I made the following changes to the code (added console logs):
In the file skill-questions-tab.directive
```
$scope.$on(EVENT_SKILL_INITIALIZED, function() {
              // eslint-disable-next-line no-console
              console.log('1');
              _init();
            });
```
In the file skill-navbar.directive
```
$scope.$on(EVENT_SKILL_INITIALIZED, function() {
              // eslint-disable-next-line no-console
              console.log('2');
              _validateSkill();
            });
```
![image](https://user-images.githubusercontent.com/23002114/79952981-d2844280-8498-11ea-9a1f-73b895bfec22.png)
Only 2 gets logged showing that questions-tab is not loaded even though it is imported.

If you would move these two services both 1 and 2 will be logged. Which is not the behavior we want.

seanlip commented:
Thanks for explaining; I understand your point about ng-if.

For the behaviour we ultimately want, see my previous comment. Again, this isn't a cookie-cutter refactor -- we need to understand why the message is being passed, and it's just to ensure that the directive has the correct skill ID value. So any approach that achieves that is fine and I am a little hesitant to call the alternative approach "wrong".

srijanreddy98 commented:
> For the behaviour we ultimately want, see my previous comment. Again, this isn't a cookie-cutter refactor -- we need to understand why the message is being passed, and it's just to ensure that the directive has the correct skill ID value. So any approach that achieves that is fine and I am a little hesitant to call the alternative approach "wrong".

There is a way of removing the $scope without any refactoring at all. This is just to detach/ decouple the refactoring from migration and taking it into a separate issue. We can have both of them going on in parallel.
@seanlip can you have a look at this?
https://github.com/oppia/oppia/pull/9136

srijanreddy98 committed:
Loader Service

srijanreddy98 committed:
Loader service

srijanreddy98 committed:
Loader service

srijanreddy98 committed:
Merge branch 'develop' into migrating-rootscope

srijanreddy98 commented:
@seanlip @bansalnitish  @vojtechjelinek, I have removed the common-events.service. Instead, I have made a loader service. This is to show a loading screen. This is will be useful for migrating $rootScope.loadingMessage. Could you all ptal.

seanlip commented on core/templates/base-components/warning-loader.directive.ts:
```diff
@@ -26,11 +27,19 @@ angular.module('oppia').directive('warningLoader', [
           ctrl.$onInit = function() {
             ctrl.AlertsService = AlertsService;
+            /**
+             * TODO(@srijanreddy98): when migrating to angular 8
```
Note that this is not the correct syntax for TODOs. Use an issue number instead.

Also, I don't understand why this code is sitting in the "warning loader". That seems wrong, conceptually? The loading message is not a warning.

srijanreddy98 replied:
> Note that this is not the correct syntax for TODOs. Use an issue number instead.

Should I create an issue for this?

> Also, I don't understand why this code is sitting in the "warning loader". That seems wrong, conceptually? The loading message is not a warning.

True, But this is where the loading message is shown. `warning-loader.directive.html` line number 21.

srijanreddy98 replied:
Also in this particular case, I haven't found a pattern like the other TODOs in the changes, should I still create an issue for this?

seanlip replied:
Oh, I see -- yes, you're right. Do you think it might be worth separating into warning-and-alerts.directive.html and loading-message.directive.html since they're pretty different? I think you can make that call.

> Also in this particular case, I haven't found a pattern like the other TODOs in the changes, should I still create an issue for this?

Sorry, I don't understand -- what do you mean by "I haven't found a pattern like the other TODOs in the changes"?

srijanreddy98 replied:
I meant like the other places where I replaced TODO's with an issue number. Issues #9154 and #9155 were created for a large range of code changes that might happen due to migration. This one, however, seems to be all alone in a league of its own.

seanlip replied:
Ah, well -- there are really only two cases for handling TODOs.

- Either (preferably): do it in the PR itself and don't have a TODO at all.
- Or, if that's not possible, file an issue and have the TODO link to the issue. If it makes sense, assign the issue to yourself for follow-up at the appropriate time.

These are the only two cases. We've found that "TODO(author)" comments in the codebase simply don't get resolved, after years.

seanlip commented on core/templates/components/entity-creation-services/skill-creation.service.ts:
```diff
@@ -16,49 +16,56 @@
+    this.skillCreationBackendApiService.createSkill(
+      description, rubrics, explanation, linkedTopicIds
+    ).then((response: SkillCreationResponse) => {
+      // TODO(srijanreddy98): Replace window with angular router
```
Same note about TODOs.

seanlip commented on core/templates/components/entity-creation-services/skill-creation.service.ts:
```diff
@@ -16,49 +16,56 @@
+      description, rubrics, explanation, linkedTopicIds
+    ).then((response: SkillCreationResponse) => {
+      // TODO(srijanreddy98): Replace window with angular router
+      window.location.href = this.urlInterpolationService.interpolateUrl(
```
Note that the previous code opens in a new tab.

srijanreddy98 replied:
I checked it, didn't open a new tab in my case. I think in order to open in a new tab, `window.open` is used. I cross-checked with develop branch of my fork, for me it didn't open a new tab. However if opening in a new tab is desired functionality, I will change it to `window.open`.

seanlip replied:
In the original code, _blank should open a new window/tab. See https://developer.mozilla.org/en-US/docs/Web/API/Window/open

seanlip commented on core/templates/services/loader.service.ts:
```diff
@@ -0,0 +1,53 @@
+  constructor() {}
+
+  setLoadingMessage(message: string): void {
+    // TODO(srijanreddy98): Change LoaderService to this.
```
Put quotes around "this", otherwise it's hard to pass.

What's preventing the use of "this" now?

srijanreddy98 replied:
Oh, because loadingMessageSubject is a static variable we can't use "this" to refer to it. Reasons for it being static are mentioned in a comment where the variable is declared.

seanlip replied:
Ah, ok, you might want to mention that in this comment then.

seanlip reviewed (changes requested):
Hi @srijanreddy98, thanks for the PR. I don't really see many issues code-wise. Just wanted to check a couple of things:

(a) have you tested this manually and is the behaviour the same in both develop and this branch? I caught one case where I believe the open-in-new-tab functionality got changed, which worries me. Please enumerate the manual tests you did for both the loading message and the skill creation service.

(b) It is weird to combine two completely disparate things in the same PR. I suggest that you create a separate PR for each one.

srijanreddy98 committed:
Code review

srijanreddy98 committed:
Merge branch 'migrating-rootscope' of https://github.com/srijanreddy98/oppia into migrating-rootscope

srijanreddy98 commented:
@seanlip I have addressed your comments (inline as well), could you ptal?
> (a) have you tested this manually and is the behavior the same in both develop and this branch? I caught one case where I believe the open-in-new-tab functionality got changed, which worries me. Please enumerate the manual tests you did for both the loading message and the skill creation service.

I did check for it, it didn't seem to open in a new tab for me. I will still recheck everything.

> (b) It is weird to combine two completely disparate things in the same PR. I suggest that you create a separate PR for each one.

Separate as in one for loader-service and the other for Migrate skill creation service to angular?

seanlip reviewed (commented):
Thanks, @srijanreddy98! Responses inline.

 >Separate as in one for loader-service and the other for Migrate skill creation service to angular? 

Yep, that's right! They are completely different things.

nithusha21 commented:
Hi @srijanreddy98  could you please merge with develop as there were some new checks introduced, and a change to the development workflow which will help make sure develop will not break after merging this PR.

Showtim3 commented:
Hey @srijanreddy98 Please resolve the conflicts.

srijanreddy98 commented:
Closing this pr for now.

This pr did two things:
1.  Create a loader service to remove `$rootScope.loadingMeaage`.
2. Migrate skill creation service.

Loader Service was created and merged in #9156.
Migrating skill creation is not possible until `EVENT_TOPICS_AND_SKILLS_DASHBOARD_REINITIALIZED` has been migrated.
